sequoia/methods/d3rlpy_methods/base.py

```diff
@@ -0,0 +1,153 @@
+from typing import Type, ClassVar, List, Tuple, Dict, Optional, Union
+
+import gym
+from d3rlpy.algos import *
+import numpy as np
+from gym import Space
+from gym.wrappers.record_episode_statistics import RecordEpisodeStatistics
+from sequoia import Method, Environment, Observations, Actions, Rewards
+from sequoia.settings.offline_rl.setting import OfflineRLSetting
+
+
+class OfflineRLWrapper(gym.Wrapper):
+    def __init__(self, env):
+        super().__init__(env)
+        self.observation_space = env.observation_space.x
+
+    def reset(self):
+        observation = super().reset()
+        return observation.x
+
+    def step(self, action):
+        observation, reward, done, info = super().step(action)
+        return observation.x, reward.y, done, info
+
+
+class BaseOfflineRLMethod(Method, target_setting=OfflineRLSetting):
+    Algo: ClassVar[Type[AlgoBase]] = AlgoBase
+
+    def __init__(self, train_steps: int = 1_000_000,
+                 train_steps_per_epoch=1_000_000,
+                 test_steps=1_000,
+                 scorers: dict = None,
+                 use_gpu: bool = False):
+        super().__init__()
+        self.train_steps = train_steps
+        self.train_steps_per_epoch = train_steps_per_epoch
+        self.test_steps = test_steps
+        self.scorers = scorers
+
+        # TODO: does use_gpu here actually work?
+        self.algo = type(self).Algo(use_gpu=use_gpu)
+
+    def configure(self, setting: OfflineRLSetting) -> None:
+        super().configure(setting)
+        self.setting = setting
+
+    def fit(self, train_env, valid_env) -> Union[None, List[Tuple[int, Dict[str, float]]]]:
```

Comment:
Why does `fit` return something?
Is this only used for testing?

Reply:
List[Tuple[int, Dict[str, float]]]] is  offline_rl results, it contains the metrics passed in to the method as a scorers dict (for example temporal difference error), as well as whatever default metrics are used (some algorithms record actor loss, critic loss, etc.. These vary completely from algorithm to algorithm). Should I not be returning this from fit?

Reply:
I see. That's not quite what I meant by a "Results" object. This is related to the other comment I left below about how you implemented `apply` in your `OfflineRLSetting`.

About the scorers, there's an issue here:
- Settings shouldn't assume anything about how the method is implemented. It should just know that the method can be trained on a given environment/dataset, and that it can be queried for actions using `get_actions`. Your setting shouldn't assume that methods are written using `d3rlpy`, or that they can have a `scorers` passed to a `fit` or `fit_online` to get back the statistics.
  
The setting should give back a `Results` object, no matter if the method is created using `d3rlpy`, or is just a random baseline method (`RandomBaselineMethod`, more specifically).

That's why I was mentioning the `RecordEpisodeStatisticsWrapper`, and I was showing you some `OfflineRLResults` dataclasses in slack. The point is: You have to do the "test loop" inside the setting, and query the method for actions. Then, whatever metrics are relevant in the setting get stored inside a `Results` object, in this case, something like `OfflineRLResults`, which you need to add.

Does that make sense?

Reply:
Added OfflineRL results. Check out how I save the offline metrics btw. It doesn't assume anything about the method, just accepts them and saves them in the result object... they very well could just be None.

Reply:
`fit` shouldn't return anything. The only reason why it could return something is if that were used in some unit tests or something, since the setting should just ignore whatever is returned by `fit`.

Reply:
Oh, I might have read that too fast, they are allowed to be None, I see.
Hmm. Then I'd say, it's not super clear why they would be in the results for some methods, but not for others. I'll have to think about this a bit.

For now, keep them as optional.

Reply:
I actually just removed these metrics and just added them as a var for BaseOfflineRLMethod

Reply:
Let me know if that looks good for you!